src/particles.cpp

```diff
@@ -18,7 +18,58 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 */
 
 #include "particles.h"
-#include "util/serialize.h"
+using namespace ParticleParamTypes;
+
+v2f ParticleParamTypes::vectorBlend(float* f, const v2f& a, const v2f& b) {
+	return v2f(
+		numericalBlend(f[0], a.X, b.X),
+		numericalBlend(f[1], a.Y, b.Y)
+	);
+}
+v3f ParticleParamTypes::vectorBlend(float* f, const v3f& a, const v3f& b) {
+	return v3f(
+		numericalBlend(f[0], a.X, b.X),
+		numericalBlend(f[1], a.Y, b.Y),
+		numericalBlend(f[2], a.Z, b.Z)
+	);
+}
+
+#define PARAM_DEF_SRZR(T, wr, rd) \
+	void ParticleParamTypes::serializeParameterValue  (std::ostream& os, T  v) {wr(os,v);  } \
+	void ParticleParamTypes::deSerializeParameterValue(std::istream& is, T& v) {v = rd(is);}
+
+PARAM_DEF_SRZR(u8,  writeU8,    readU8);
+PARAM_DEF_SRZR(u16, writeU16,   readU16);
+PARAM_DEF_SRZR(u32, writeU32,   readU32);
+PARAM_DEF_SRZR(f32, writeF32,   readF32);
+PARAM_DEF_SRZR(v2f, writeV2F32, readV2F32);
+PARAM_DEF_SRZR(v3f, writeV3F32, readV3F32);
+
+#undef PARAM_DEF_SRZR
+
+void ServerParticleTexture::serialize(std::ostream &os, u16 protocol_ver) const {
+	u8 flags = 0;
+	animated && (flags |= 1<<0);
+	writeU8(os, flags);
+
+	alpha.serialize(os);
+	scale.serialize(os);
+	os << serializeString32(string);
+
+	if (animated)
+		animation.serialize(os, protocol_ver);
+}
+void ServerParticleTexture::deSerialize(std::istream &is, u16 protocol_ver) {
+	u8 flags = readU8(is);
+	animated = (flags |= 1<<0) > 0;
```

Comment:
```suggestion
	animated = !!(flags & 1);
```
also wouldn't hurt to put them in an enum or #define

Reply:
whooooops. good catch ty fixed

i didn't bother adding names to the flags simply because there's only one flag right now and they will only ever be used in two specific places, but if it'll make the protocol easier to make sense of i'll move the magic numbers into an enum